tests/unit_tests/test_heron.py

```diff
@@ -1,61 +1,669 @@
+        # Verify CashFlow with type
+        cashflows = comp.findall('./economics/CashFlow')
+        self.assertEqual(len(cashflows), 1)
+        self.assertEqual(cashflows[0].attrib['type'], 'one-time')
+      elif comp.get('name') == 'ExistingComponent1':
+        # Verify CashFlow with type
+        cashflows = comp.findall('./economics/CashFlow')
         self.assertEqual(len(cashflows), 1)
-        self.assertEqual(cashflows[0].attrib['name'], 'NewComponent_capex')
+        self.assertEqual(cashflows[0].attrib['type'], 'repeating')
+      elif comp.get('name') == 'NewComponent':
+        # Verify reference driver
+        ref_driver = comp.find('./economics/CashFlow/reference_driver/fixed_value')
+        self.assertEqual(ref_driver.text, '1000') # Check that mW were not converted
+
+    # Verify DataGenerators node was transferred
+    with (self.subTest("DataGenerators node has been corrupted")):
+      data_gens = result_tree.findall('./DataGenerators')
+      self.assertEqual(len(data_gens), 1)
+      self.assertIsNotNone(data_gens[0].find('./untouched_content_DG'))
+
+class TestExpandedInput2(unittest.TestCase):
+  
+  def setUp(self):
+    # Complex subnodes to each component with various <economics> and <CashFlows> positions and configurations
+    self.heron_xml = """<HERON>
+                          <Components>
+                            <Component name="Component0">
+                              <economics>
+                                <CashFlow name="other">
+                                  <untouched_content></untouched_content>
+                                </CashFlow>
+                              </economics>
+                            </Component>
+
+                            <Component name="Component1">
+                              <produces resource="something">
+                                <some_other_info></some_other_info>
+                              </produces>
+                              <economics>
+                                <CashFlow name="other" type="one-time">
+                                </CashFlow>
+                                <CashFlow name="capex" type="one-time">
+                                </CashFlow>
+                              </economics>
+                            </Component>
+
+                            <Component name="Component2">
+                              <economics>
+                                <ProjectTime>1</ProjectTime>
+                                <CashFlow name="capex_Comp2" npv_exempt="True">
+                                  <reference_driver>
+                                    <fixed_value>1234</fixed_value>
+                                  </reference_driver>
+                                  <reference_price>
+                                    <ROM rom_stuff="stuff">more_rom_stuff</ROM>
+                                  </reference_price>
+                                  <scaling_factor_x>
```

Comment:
I think it's good to check these mixed-ValuedParam cases. From what I'm seeing here, every child node here gets replaced with a value from the `mock_open` callable below. Now this might be a question of ignorance, but: would it be possible to leave one of these parameters unchanged? By that I mean, for example, that the end goal for `Component 2` is to replace the `reference_driver`, the `reference_price`, and the `driver` with values read in from Line 246-248, BUT we wanted to keep the `scaling_factor_x` as an uncertainty. Is that possible to test for/a current capability of the method?

Reply:
The reason I chose not to modify the set of ValuedParams is that, in the current functionality of the `src/heron.py` script, all the ValuedParam inputs from ASPEN HYSYS and APEA are required. The `src/heron.py` script currently trusts that the ASPEN HYSYS and APEA inputs have been provided, so it does not reliably throw an error when one is missing. Thus, neither a functional nor an erroneous response to the omission of some of the ValuedParams can be tested.